test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+            await self.client.admin.command(
+                "configureFailPoint", "maxTimeAlwaysTimeOut", mode="off"
+            )
+
+    def test_with_options(self):
+        codec_options = DECIMAL_CODECOPTS
+        read_preference = ReadPreference.SECONDARY_PREFERRED
+        write_concern = WriteConcern(j=True)
+        read_concern = ReadConcern(level="majority")
+
+        # List of all options to compare.
+        allopts = [
+            "name",
+            "client",
+            "codec_options",
+            "read_preference",
+            "write_concern",
+            "read_concern",
+        ]
+
+        db1 = self.client.get_database(
+            "with_options_test",
+            codec_options=codec_options,
+            read_preference=read_preference,
+            write_concern=write_concern,
+            read_concern=read_concern,
+        )
+
+        # Case 1: swap no options
+        db2 = db1.with_options()
+        for opt in allopts:
+            self.assertEqual(getattr(db1, opt), getattr(db2, opt))
+
+        # Case 2: swap all options
+        newopts = {
+            "codec_options": CodecOptions(),
+            "read_preference": ReadPreference.PRIMARY,
+            "write_concern": WriteConcern(w=1),
+            "read_concern": ReadConcern(level="local"),
+        }
+        db2 = db1.with_options(**newopts)  # type: ignore[arg-type]
+        for opt in newopts:
+            self.assertEqual(getattr(db2, opt), newopts.get(opt, getattr(db1, opt)))
+
+
+class TestDatabaseAggregation(AsyncIntegrationTest):
+    def setUp(self):
+        self.pipeline: List[Mapping[str, Any]] = [
+            {"$listLocalSessions": {}},
+            {"$limit": 1},
+            {"$addFields": {"dummy": "dummy field"}},
+            {"$project": {"_id": 0, "dummy": 1}},
+        ]
+        self.result = {"dummy": "dummy field"}
+        self.admin = self.client.admin
+
+    async def test_database_aggregation(self):
+        cursor = await self.admin.aggregate(self.pipeline)
+        result = await anext(cursor)
+        self.assertEqual(result, self.result)
+
+    @async_client_context.require_no_mongos
+    async def test_database_aggregation_fake_cursor(self):
+        coll_name = "test_output"
+        write_stage: dict
+        if async_client_context.version < (4, 3):
+            db_name = "admin"
+            write_stage = {"$out": coll_name}
+        else:
+            # SERVER-43287 disallows writing with $out to the admin db, use
+            # $merge instead.
+            db_name = "pymongo_test"
+            write_stage = {"$merge": {"into": {"db": db_name, "coll": coll_name}}}
+        output_coll = self.client[db_name][coll_name]
+        await output_coll.drop()
+        self.addAsyncCleanup(output_coll.drop)
+
+        admin = self.admin.with_options(write_concern=WriteConcern(w=0))
+        pipeline = self.pipeline[:]
+        pipeline.append(write_stage)
+        cursor = await admin.aggregate(pipeline)
```

Comment:
You can use `async with await self.admin.aggregate(self.pipeline) as cursor:` and keep it similar to how the synchronous version was before.